# Домашнее задание 1
В этом задании мы:
1. Вспомним Python: циклы, функции и рекурсию.
2. Посоздаем тензоры, поиндексируем их.
3. Вручную имплементируем One-Hot Encoder.

## Вспоминаем Python
Перед тем, как нырнуть в глубинное обучение, вспомним азы: язык Python.
Нам понадобится работать со списками и вложенными вызовами функций, попрактикуемся с этим.

### Задание №1:

Напишите функцию, которая принимает на вход массив и возвращает другой массив, состоящий только из уникальных элементов исходного массива.
Проще говоря, функция должна вернуть уникальные элементы из исходного массива.

Функция должна называться `unique_elements`. На вход функции будет передаваться стандартный тип `list` из Python. На выходе функции ожидается стандартный тип `list` из Python.

In [1]:
def unique_elements(list_: list[int]) -> list[int]:
    
    return list(set(list_))

### Задание №2:
Напишите функцию для подсчета n-го числа Фиббоначчи.

P.S формула n-го числа Фиббоначчи:
\begin{cases}
F(0) = 0 \\
F(1) = 1 \\
F(n) = F(n-1) + F(n-2), \quad \text{для } n \geq 2
\end{cases}


In [27]:
def fib(n: int) -> int:
    
    fib_n = 0
    if n >= 2:
        fib_list = [0, 1]
        for i in range(n-1):
            sum_elem = fib_list[-1] + fib_list[-2]
            
            print(f"fib_list[-1]: {fib_list[-1]}")
            print(f"fib_list[-2]: {fib_list[-2]}")
            fib_list.append(sum_elem)
            # print(fib_list[-1] + fib_list[-2])
        fib_n = fib_list[-1]
    else:
        fib_n = n

    return fib_n

In [20]:
fib_list = [0, 1]

In [22]:
fib_list[-2]

0

In [30]:
fib(10)

fib_list[-1]: 1
fib_list[-2]: 0
fib_list[-1]: 1
fib_list[-2]: 1
fib_list[-1]: 2
fib_list[-2]: 1
fib_list[-1]: 3
fib_list[-2]: 2
fib_list[-1]: 5
fib_list[-2]: 3
fib_list[-1]: 8
fib_list[-2]: 5
fib_list[-1]: 13
fib_list[-2]: 8
fib_list[-1]: 21
fib_list[-2]: 13
fib_list[-1]: 34
fib_list[-2]: 21


55

In [3]:
list(range(2))

[0, 1]

## Тензоры

Перейдем ближе к DL. Нейросети оперируют понятием **тензора**.

Тензор - это многомерная матрица.
Попрактикуемся в работе с ними.

In [1]:
import torch
import numpy as np

In [129]:
torch.ones(size=[1000, 1000], dtype=int) * torch.arange(1000)

tensor([[  0,   1,   2,  ..., 997, 998, 999],
        [  0,   1,   2,  ..., 997, 998, 999],
        [  0,   1,   2,  ..., 997, 998, 999],
        ...,
        [  0,   1,   2,  ..., 997, 998, 999],
        [  0,   1,   2,  ..., 997, 998, 999],
        [  0,   1,   2,  ..., 997, 998, 999]])

### Задание №3:
Создайте тензор размера (5, 3), заполненный случайными целыми числами от -1 до 1.

Запишите результат в переменную t

In [4]:
t = torch.randint(-1, 2, (5, 3))

### Задание №4:
Вам дана доска 1000x1000.
Каждую строку в ней нумеруют от 0 до 999 включительно.
Т.е. получится матрица вида
```python
[
    [0, 1, ..., 999],
    [0, 1, ..., 999],
    ...
]
```
Напишите код, который сгенерирует такую матрицу. Постарайтесь обойтись без циклов `for`.
Запишите эту матрицу в переменную `result`.

_Подсказка_: посмотрите в сторону метода `some_tensor.repeat((N, M))`

In [5]:
result = torch.ones([1000, 1000]) * torch.arange(1000)

### Задание №5:
Напишите функцию, которая будет нормировать трехмерный тензор вдоль его двух последних осей.
Если на вход передается тензор размером `(5, 2, 10)`, необходимо выполнить следующие шаги:
- Для каждого из 5 срезов (плоскостей размером `2x10`) вычислить среднее значение и дисперсию всех элементов в этой плоскости.
- Используя эти средние значения и дисперсии, нормировать все элементы в соответствующей плоскости.


Приводим ниже код через циклы `for` - вам нужно написать решение без использования `for`.

In [258]:
def slow_normalize(x: torch.Tensor) -> torch.Tensor:
    for i in range(x.shape[0]):
        mean, std = x[i].mean(), x[i].std()
        x[i] = (x[i] - mean) / std
    return x


def fast_normalize(x: torch.Tensor) -> torch.Tensor:
    
    s_norm = (x - x.mean(dim=1, keepdim=True)) / x.std(dim=1, keepdim=True)

    return s_norm


from torch.testing import assert_close

data = torch.rand(5, 2, 10)
out1 = slow_normalize(data)
out2 = fast_normalize(data)
# assert_close(out2, out1)

RuntimeError: mean(): could not infer output dtype. Input dtype must be either a floating point or complex dtype. Got: Long

In [257]:
out2

tensor([[[-0.7071,  0.7071,  0.7071,  0.7071, -0.7071,  0.7071,  0.7071,
          -0.7071, -0.7071, -0.7071],
         [ 0.7071, -0.7071, -0.7071, -0.7071,  0.7071, -0.7071, -0.7071,
           0.7071,  0.7071,  0.7071]],

        [[ 0.7071,  0.7071, -0.7071,  0.7071,  0.7071, -0.7071, -0.7071,
          -0.7071, -0.7071,  0.7071],
         [-0.7071, -0.7071,  0.7071, -0.7071, -0.7071,  0.7071,  0.7071,
           0.7071,  0.7071, -0.7071]],

        [[ 0.7071,  0.7071,  0.7071,  0.7071, -0.7071,  0.7071,  0.7071,
          -0.7071,  0.7071,  0.7071],
         [-0.7071, -0.7071, -0.7071, -0.7071,  0.7071, -0.7071, -0.7071,
           0.7071, -0.7071, -0.7071]],

        [[-0.7071, -0.7071,  0.7071,  0.7071, -0.7071,  0.7071,  0.7071,
          -0.7071,  0.7071, -0.7071],
         [ 0.7071,  0.7071, -0.7071, -0.7071,  0.7071, -0.7071, -0.7071,
           0.7071, -0.7071,  0.7071]],

        [[ 0.7071, -0.7071,  0.7071, -0.7071, -0.7071, -0.7071, -0.7071,
           0.7071, -0.7071,  0.

In [157]:
(torch.arange(5 * 4 * 3, dtype=torch.float32).reshape((5, 4, 3)) ** 2).shape[0]

5

In [194]:
torch.linspace(0, 15, 15)

tensor([ 0.0000,  1.0714,  2.1429,  3.2143,  4.2857,  5.3571,  6.4286,  7.5000,
         8.5714,  9.6429, 10.7143, 11.7857, 12.8571, 13.9286, 15.0000])

In [231]:
a = torch.ones(5, 2, 10)

In [269]:
a

tensor([[[   1.,    1.,    1.,    1.,    1.,    1.,    1.,    1.,    1.,    1.],
         [   1.,    1.,    1.,    1.,    1.,    1.,    1.,    1.,    1.,    1.]],

        [[   2.,    2.,    2.,    2.,    2.,    2.,    2.,    2.,    2.,    2.],
         [   2.,    2.,    2.,    2.,    2.,    2.,    2.,    2.,    2.,    2.]],

        [[-100.,   40.,   40.,   40.,   40.,   40.,   40.,   40.,   40.,   40.],
         [   4.,    4.,    4.,    4.,    4.,    4.,    4.,    4.,    4.,    4.]],

        [[   8.,    8.,    8.,    8.,    8.,    8.,    8.,    8.,    8.,    8.],
         [   8.,    8.,    8.,    8.,    8.,    8.,    8.,    8.,    8.,    8.]],

        [[   0.,    0.,    0.,    0.,    0.,    0.,    0.,    0.,    0.,    0.],
         [ 100.,  100.,  100.,  100.,  100.,  100.,  100.,  100.,  100.,  100.]]])

In [268]:
a[2][0][0] = -100

In [272]:
a.mean(dim=0, keepdim=True)

tensor([[[-17.8000,  10.2000,  10.2000,  10.2000,  10.2000,  10.2000,  10.2000,
           10.2000,  10.2000,  10.2000],
         [ 23.0000,  23.0000,  23.0000,  23.0000,  23.0000,  23.0000,  23.0000,
           23.0000,  23.0000,  23.0000]]])

In [270]:
a.mean(dim=2)

tensor([[  1.,   1.],
        [  2.,   2.],
        [ 26.,   4.],
        [  8.,   8.],
        [  0., 100.]])

In [263]:
a.mean(dim=(1, 2), keepdim=True)

tensor([[[ 1.]],

        [[ 2.]],

        [[ 4.]],

        [[ 8.]],

        [[50.]]])

In [167]:
slow_normalize(torch.Tensor(5, 2, 10))

tensor([[[-1.1325e+00, -1.1325e+00, -1.0853e+00,  7.2707e-01, -3.8232e-04,
           1.6779e+00, -2.7253e-01, -3.0278e-01,  1.1623e+00,  7.4044e-01],
         [-8.5626e-01, -8.7998e-01, -5.8369e-03, -6.5061e-01, -5.2264e-01,
           2.4790e+00, -4.9970e-01,  1.0396e+00, -8.4499e-02, -4.0075e-01]],

        [[ 3.3708e-01, -9.8784e-01, -1.4087e+00,  2.9703e-01, -9.0077e-01,
          -1.0392e+00, -1.0485e+00, -3.9487e-01,  1.2986e+00, -8.6167e-01],
         [-5.0440e-01, -5.6060e-01,  1.7222e+00,  1.4355e+00,  3.1048e-01,
           1.6770e+00,  2.9702e-01,  2.6257e-01,  9.2858e-01, -8.5951e-01]],

        [[-7.8473e-01, -1.5916e+00,  1.0419e+00, -7.6645e-01,  6.9101e-01,
           8.8511e-01,  1.3650e+00, -2.5212e-02, -8.5524e-01, -8.5813e-01],
         [-1.1403e+00, -7.6061e-01,  4.7438e-01,  7.6070e-01, -3.6614e-01,
          -1.4634e+00,  6.2056e-01,  1.7492e+00, -6.1732e-02,  1.0857e+00]],

        [[ 1.8690e-01,  1.6028e-01, -1.1050e+00, -1.5205e+00,  1.2317e+00,
          -1.

In [154]:
torch.arange(5 * 4 * 3, dtype=torch.float32).reshape((5, 4, 3)) ** 2

tensor([[[0.0000e+00, 1.0000e+00, 4.0000e+00],
         [9.0000e+00, 1.6000e+01, 2.5000e+01],
         [3.6000e+01, 4.9000e+01, 6.4000e+01],
         [8.1000e+01, 1.0000e+02, 1.2100e+02]],

        [[1.4400e+02, 1.6900e+02, 1.9600e+02],
         [2.2500e+02, 2.5600e+02, 2.8900e+02],
         [3.2400e+02, 3.6100e+02, 4.0000e+02],
         [4.4100e+02, 4.8400e+02, 5.2900e+02]],

        [[5.7600e+02, 6.2500e+02, 6.7600e+02],
         [7.2900e+02, 7.8400e+02, 8.4100e+02],
         [9.0000e+02, 9.6100e+02, 1.0240e+03],
         [1.0890e+03, 1.1560e+03, 1.2250e+03]],

        [[1.2960e+03, 1.3690e+03, 1.4440e+03],
         [1.5210e+03, 1.6000e+03, 1.6810e+03],
         [1.7640e+03, 1.8490e+03, 1.9360e+03],
         [2.0250e+03, 2.1160e+03, 2.2090e+03]],

        [[2.3040e+03, 2.4010e+03, 2.5000e+03],
         [2.6010e+03, 2.7040e+03, 2.8090e+03],
         [2.9160e+03, 3.0250e+03, 3.1360e+03],
         [3.2490e+03, 3.3640e+03, 3.4810e+03]]])

In [136]:
torch.arange(5 * 4 * 3, dtype=torch.float32).reshape((5, 4, 3)) ** 2

tensor([[[0.0000e+00, 1.0000e+00, 4.0000e+00],
         [9.0000e+00, 1.6000e+01, 2.5000e+01],
         [3.6000e+01, 4.9000e+01, 6.4000e+01],
         [8.1000e+01, 1.0000e+02, 1.2100e+02]],

        [[1.4400e+02, 1.6900e+02, 1.9600e+02],
         [2.2500e+02, 2.5600e+02, 2.8900e+02],
         [3.2400e+02, 3.6100e+02, 4.0000e+02],
         [4.4100e+02, 4.8400e+02, 5.2900e+02]],

        [[5.7600e+02, 6.2500e+02, 6.7600e+02],
         [7.2900e+02, 7.8400e+02, 8.4100e+02],
         [9.0000e+02, 9.6100e+02, 1.0240e+03],
         [1.0890e+03, 1.1560e+03, 1.2250e+03]],

        [[1.2960e+03, 1.3690e+03, 1.4440e+03],
         [1.5210e+03, 1.6000e+03, 1.6810e+03],
         [1.7640e+03, 1.8490e+03, 1.9360e+03],
         [2.0250e+03, 2.1160e+03, 2.2090e+03]],

        [[2.3040e+03, 2.4010e+03, 2.5000e+03],
         [2.6010e+03, 2.7040e+03, 2.8090e+03],
         [2.9160e+03, 3.0250e+03, 3.1360e+03],
         [3.2490e+03, 3.3640e+03, 3.4810e+03]]])

In [150]:
torch.Tensor(5, 2, 10)

tensor([[[-3.1001e-06,  1.2107e-42, -2.0000e+00,  1.8750e+00,  3.6893e+19,
           1.8568e+00,  3.6893e+19,  1.8264e+00, -1.0842e-19,  1.7839e+00],
         [ 2.0000e+00,  1.7085e+00,  2.0000e+00,  1.5248e+00,  1.0842e-19,
           1.5410e+00,  1.0842e-19,  1.7576e+00, -2.0000e+00,  1.8609e+00]],

        [[-1.0842e-19,  1.9256e+00, -1.0842e-19,  1.9894e+00,  0.0000e+00,
           1.9219e+00,  3.6893e+19,  1.8972e+00, -1.0842e-19,  1.8661e+00],
         [-0.0000e+00,  1.8089e+00, -2.0000e+00,  1.7454e+00,  0.0000e+00,
           1.6052e+00, -1.0842e-19,  1.3092e+00, -3.6893e+19,  1.6691e+00]],

        [[ 0.0000e+00,  1.7865e+00,  0.0000e+00,  1.8674e+00,  0.0000e+00,
           1.9136e+00, -0.0000e+00,  1.9580e+00,  0.0000e+00,  1.9297e+00],
         [-0.0000e+00,  1.9009e+00, -0.0000e+00,  1.8670e+00, -0.0000e+00,
           1.8048e+00, -0.0000e+00,  1.7304e+00,  0.0000e+00,  1.5683e+00]],

        [[ 1.0842e-19,  1.4174e+00,  3.6893e+19,  1.6850e+00,  1.0842e-19,
           1.

### Задание №6:
Вам дан тензор `t` размерности (53, 192, 789), в котором зашит секретный пароль.
Пароль хитро разбросан по тензору, но вы знаете, как его отыскать:
- вдоль первой размерности нужно взять индексы, являющиеся точными квадратами
- вдоль второй размерности нужно взять столько четных чисел (начиная с 0), сколько было в п.1;
- вдоль третьей размерности нужно взять в обратном порядке столько элементов с конца, сколько было в п.2;

Найдите тензор с паролем. Исходный тензор лежит в переменной `t`.

Запишите тензор с паролем в переменную `password`.

Для тестирования можете создать тензор так:
```python
torch.random.manual_seed(0)
t = torch.randint(0, 100, (53, 192, 789))
```

_Подсказка_: если PyTorch не хочет разворачивать, попробуйте `.flip()`

_Подсказка_: индексы, которые хотите забрать, можно положить в список - и передать список в виде `t[my_list]`

P.S Точный квадрат - число, являющееся квадратом некоторого целого числа. Иными словами, квадратом является целое число, квадратный корень из которого извлекается нацело.

In [7]:
torch.random.manual_seed(0)
t = torch.randint(0, 100, (53, 192, 789))


def password(t: torch.Tensor) -> torch.Tensor:
    return ...

In [17]:
import torch

def extract_password(t):
    square_indices = [n**2 for n in range(int(53**0.5) + 1) if n**2 < 53]
    
    even_indices = list(range(0, 2 * len(square_indices), 2))
    
    reverse_slice = torch.arange(789 - 1, 789 - len(even_indices) - 1, -1)
    
    password_tensor = t[square_indices, :, :][:, even_indices, :][:, :, reverse_slice]
    
    return password_tensor

torch.Size([8, 8, 8])
tensor([[[81, 40, 56, 69,  2, 65, 54, 84],
         [77, 73, 57, 45, 75, 90,  2,  1],
         [ 3, 47, 74, 74, 58, 79, 85, 90],
         [33, 55, 40, 91, 41, 42, 97,  6],
         [59, 79, 59, 57,  0, 43, 21,  3],
         [31, 18, 14, 60, 89, 36, 50, 33],
         [17, 61, 77, 17, 57, 26, 66, 89],
         [43, 77, 85, 74, 33, 21, 61, 73]],

        [[16, 69,  1, 79, 64, 24, 14, 54],
         [47, 69,  7,  6, 72, 39, 69,  4],
         [86,  0, 68,  0, 36,  9,  8, 65],
         [74, 83, 30, 40, 27, 23, 54, 96],
         [10, 61, 83, 55, 13,  4, 39, 58],
         [82, 28, 97, 59, 50, 58, 14, 36],
         [26, 39, 87, 76,  3, 85,  0, 43],
         [74,  8, 40, 61, 29, 51, 72, 59]],

        [[38,  4, 26, 48,  0, 13, 50, 72],
         [37, 73, 81, 80,  3, 32, 76, 83],
         [54, 33, 91, 49,  7, 68, 52, 84],
         [95, 59, 53, 16, 47, 65, 21, 36],
         [56, 83, 29, 35, 33, 93, 59, 87],
         [12, 96, 50, 30, 91, 42, 75, 79],
         [27, 18, 80, 23, 33

In [33]:
import torch
torch.random.manual_seed(0)
t = torch.randint(0, 100, (53, 192, 789))

square_indices = [n**2 for n in range(int(53**0.5) + 1) if n**2 < 53]

even_indices = list(range(0, 2 * len(square_indices), 2)) 

reverse_slice = torch.arange(789 - 1, 789 - len(even_indices) - 1, -1)

password = t[square_indices, :, :][:, even_indices, :][:, :, reverse_slice]

In [37]:
list(range(0, 20 * 2, 2))

[0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38]

In [18]:
import torch
torch.random.manual_seed(0)
t = torch.randint(0, 100, (53, 192, 789))
square_indices = [n**2 for n in range(int(53**0.5) + 1) if n**2 < 53]
even_indices = list(range(0, len(square_indices) + 1, 2)) 
reverse_slice = torch.arange(789 - 1, 789 - len(even_indices) - 1, -1)
password = t[square_indices, :, :][:, even_indices, :][:, :, reverse_slice]

password = password.flatten()


In [16]:
# Пример с тестовым тензором
torch.random.manual_seed(0)
t = torch.randint(0, 100, (53, 192, 789))

# После выполнения кода выше:
print(password_tensor.shape)  # Должно быть torch.Size([8, 8, 8])
print(password[:10])  # Пример вывода: tensor([40, 49, 83, ..., ...])

torch.Size([8, 8, 8])
tensor([81, 40, 56, 69,  2, 65, 54, 84, 77, 73])


In [20]:
import torch

torch.manual_seed(0)
t = torch.randint(0, 100, (53, 192, 789))

idx0 = [i*i for i in range(8) if i*i < 53]

n = len(idx0)  # n = 8

idx1 = list(range(0, 2*n, 2))

idx2 = list(range(t.shape[2] - 1, t.shape[2] - n - 1, -1))

password = t[idx0][:, idx1][:, :, idx2]


In [23]:
import torch

torch.manual_seed(0)
t = torch.randint(0, 100, (53, 192, 789))

idx0 = [i * i for i in range(8) if i * i < t.shape[0]]
n = len(idx0)

idx1 = [i for i in range(2 * n) if i % 2 == 0]

idx2 = list(range(t.shape[2] - 1, t.shape[2] - n - 1, -1))

idx0_tensor = torch.tensor(idx0, dtype=torch.long)
idx1_tensor = torch.tensor(idx1, dtype=torch.long)
idx2_tensor = torch.tensor(idx2, dtype=torch.long)

password = t[idx0_tensor[:, None, None],
             idx1_tensor[None, :, None],
             idx2_tensor[None, None, :]]

In [47]:
import torch

# # Устанавливаем seed для воспроизводимости
# torch.manual_seed(42)
# t = torch.randint(0, 100, (53, 192, 789))

# # Формируем список индексов для первой размерности – индексы точных квадратов
# idx0 = [i * i for i in range(8) if i * i < t.shape[0]]
# n = len(idx0)

# # Для второй размерности – первые n чётных чисел (начиная с 0)
# idx1 = [i for i in range(2 * n) if i % 2 == 0]

# # Для третьей размерности – последние n элементов в обратном порядке
# idx2 = list(range(t.shape[2] - 1, t.shape[2] - n - 1, -1))

# # Преобразуем индексы в тензоры и корректно "разворачиваем" для расширения размерностей
# i0 = torch.tensor(idx0, dtype=torch.long).view(-1, 1, 1)
# i1 = torch.tensor(idx1, dtype=torch.long).view(1, -1, 1)
# i2 = torch.tensor(idx2, dtype=torch.long).view(1, 1, -1)

# # Используем продвинутое индексирование с broadcasting
# password = t[i0, i1, i2]

In [30]:
password

tensor([[[55, 67, 62, 64, 21,  3, 39, 70],
         [37, 18, 63, 35, 90, 69, 67,  8],
         [88, 33, 13, 21, 32, 74, 96,  2],
         [64,  9, 44, 69, 17, 17, 40, 62],
         [43, 73,  7, 98, 10, 34, 17, 53],
         [80, 65, 51, 21, 54, 46, 86, 93],
         [20, 38, 84, 25, 41,  5, 56, 15],
         [73, 94, 71, 77, 58, 30, 45, 50]],

        [[34, 26, 61,  4, 15, 52, 28, 51],
         [41, 25, 30, 37, 14, 38, 86, 33],
         [22, 95,  0, 28, 75, 80, 50, 20],
         [31, 86, 94,  6, 61, 43, 54,  5],
         [ 9, 27, 63, 63,  8, 70,  5, 64],
         [77, 41, 29, 76, 24, 58, 17, 60],
         [96, 35, 35, 69,  7, 97, 72, 45],
         [12, 33, 22,  3, 67, 93, 23, 54]],

        [[89, 11, 69, 78, 79, 64, 19, 47],
         [53, 38, 74, 66, 17, 26, 80, 35],
         [86, 22, 75, 57, 44, 10, 28, 86],
         [28, 54, 45,  3,  0, 36, 24, 65],
         [33, 11, 33, 48, 68, 24, 19, 27],
         [78, 99, 50, 25, 98, 74, 79, 21],
         [86, 63, 98,  5,  4, 67,  1, 19],
       

In [3]:
a = torch.randint(0, 100, (2, 3, 4))

In [12]:
a[:, :, :]

tensor([[[96,  3, 54, 25],
         [ 7, 64, 76, 11],
         [96, 56, 50, 32]],

        [[79, 89, 76, 41],
         [75, 26, 76, 78],
         [13, 17, 26, 94]]])

In [ ]:
a[:, :, :]

tensor([[96,  7, 96],
        [79, 75, 13]])

In [4]:
torch.sqrt(a)

tensor([[[9.7980, 1.7321, 7.3485, 5.0000],
         [2.6458, 8.0000, 8.7178, 3.3166],
         [9.7980, 7.4833, 7.0711, 5.6569]],

        [[8.8882, 9.4340, 8.7178, 6.4031],
         [8.6603, 5.0990, 8.7178, 8.8318],
         [3.6056, 4.1231, 5.0990, 9.6954]]])

In [5]:
mask = a == a.round()

In [8]:
mask

tensor([[[True, True, True, True],
         [True, True, True, True],
         [True, True, True, True]],

        [[True, True, True, True],
         [True, True, True, True],
         [True, True, True, True]]])

In [7]:
a[mask]

tensor([96,  3, 54, 25,  7, 64, 76, 11, 96, 56, 50, 32, 79, 89, 76, 41, 75, 26,
        76, 78, 13, 17, 26, 94])

In [10]:
4.3 % 1

0.2999999999999998

In [1]:
2 * 2

4

### Задание №7:
Реализуйте One Hot Encoder. Вам необходимо написать функцию, которая выполняет One Hot Encoding для одномерного тензора.

In [48]:
def ohe(y: torch.Tensor) -> torch.Tensor:

    uniq_list = []
    for i in y:
        if i not in uniq_list:
            uniq_list.append(i)
    
    for i in range(len(uniq_list)):
        for j in range(1, len(uniq_list)):
            if uniq_list[j-1] > uniq_list[j]:
                uniq_list[j-1:j+1] = uniq_list[j], uniq_list[j-1]
    
    tensor_list = []
    for i in range(len(y)):
        for j in range(len(uniq_list)):
            class_list = list([0] * len(uniq_list))
            if y[i] == uniq_list[j]:
                class_list[j] = 1
                tensor_list.append(class_list)

    return torch.tensor(tensor_list)

In [49]:
ohe(torch.tensor([0, 1, 0, 1]))

tensor([[1, 0],
        [0, 1],
        [1, 0],
        [0, 1]])

In [50]:
ohe(torch.tensor([0, 1, 2]))

tensor([[1, 0, 0],
        [0, 1, 0],
        [0, 0, 1]])

In [51]:
ohe(torch.tensor([3, 2, 5, 9, 6, 3]))

tensor([[0, 1, 0, 0, 0],
        [1, 0, 0, 0, 0],
        [0, 0, 1, 0, 0],
        [0, 0, 0, 0, 1],
        [0, 0, 0, 1, 0],
        [0, 1, 0, 0, 0]])

In [23]:
y = torch.tensor([0, 1, 0, 1])

In [25]:
list(y)

[tensor(0), tensor(1), tensor(0), tensor(1)]

In [35]:
len(y)

4

In [ ]:
tensor_list = []
for i in range(len(y)):
    for j in range(len(uniq_list)):
        class_list = list([uniq_list[0]] * len(uniq_list))
        if y[i] == uniq_list[j]:
            class_list[j] = 1
            tensor_list.append(class_list)
    

In [36]:
list([0] * 2)

[0, 0]

In [29]:
y[0]

tensor(0)

In [26]:
uniq_list = []
for i in y:
    if i not in uniq_list:
        uniq_list.append(i)

In [27]:
uniq_list

[tensor(0), tensor(1)]

In [40]:

from torch.testing import assert_close
assert_close(ohe(torch.tensor([0, 1, 0, 1])), torch.tensor([[1, 0], [0, 1], [1, 0], [0, 1]]))
assert_close(ohe(torch.tensor([0, 1, 2])), torch.tensor([[1, 0, 0], [0, 1, 0], [0, 0, 1]]))

# :param y: массив размера (N,).
# :returns: массив размера (N, K), где K - число уникальных значений в y.
# """

In [43]:
torch.tensor([0, 1, 0, 1])

tensor([0, 1, 0, 1])

In [42]:
torch.tensor([[1, 0], [0, 1], [1, 0], [0, 1]])

tensor([[1, 0],
        [0, 1],
        [1, 0],
        [0, 1]])

In [12]:
d = {'number_0': [1, 0], 'number_1': [0, 1]}

In [2]:
import pandas as pd

In [15]:
df = pd.DataFrame(data=d)

In [20]:
df.values.tolist()

[[1, 0], [0, 1]]

[('number_0',
  0    1
  1    0
  Name: number_0, dtype: int64),
 ('number_1',
  0    0
  1    1
  Name: number_1, dtype: int64)]

In [ ]:
assert_close(torch)

In [13]:
2 * 2

4

In [9]:
3 * 1235

3705

In [3]:
def hah(a, default=False):
    if a > 1:
        return default
    return True

In [4]:
hah(5)

False

: 